docs(forms): expand e-mail validation description

gkalpak opened the pull request:
Previously, there was no documentation of what `Validators.email()` expects as a valid e-mail address, making it difficult for people to determine whether it covers their requirements or not. Even more so that the used pattern slightly deviates from the [WHATWG version](https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address). One's only option was to look at the source code and try to decipher the regexp pattern.

This PR adds a high-level description of the validator and mentions its similarity to and differences from the WHATWG version. It also adds a brief explanation of the regexp's behavior and references for more information in the source code to provide more context to maintainers/users trying to understand the implementation in the future.

Finally, this PR also refactors `EMAIL_REGEXP (without changing its behavior) to make it more closely match the format of WHATWG version, so that it is easier for people to compare it against the WHATWG one and understand the differences.

Fixes #18985, fixes #25186, closes #32747.

gkalpak committed:
docs(forms): expand e-mail validation description

Previously, there was no documentation of what `Validators.email()`
expects as a valid e-mail address, making it difficult for people to
determine whether it covers their requirements or not. Even more so that
the used pattern slightly deviates from the
[WHATWG version](https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address).
One's only option was to look at the source code and try to decipher the
regexp pattern.

This commit adds a high-level description of the validator and mentions
its similarity to and differences from the WHATWG version. It also adds
a brief explanation of the regexp's behavior and references for more
information in the source code to provide more context to
maintainers/users trying to understand the implementation in the future.

Fixes #18985
Fixes #25186
Closes #32747

gkalpak committed:
refactor(forms): refactor `Validators.email()` regexp for easier comparison with WHATWG version

As mentioned in the previous commit, the regexp used by
`Validators.email()` is a slightly enhanced version of the
[WHATWG one](https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address).

This commit refactors the regexp (without changing its behavior) to make
it more closely match the format of WHATWG version, so that it is easier
for people to compare it against the WHATWG one and understand the
differences.

The main changes were:
- Changing the order of characters/character classes inside `[...]`;
  e.g. `[A-Za-z]` --> `[a-zA-Z]`
- Mark all groups as non-capturing (since we do not use the captured
  values); e.g. `(foo)` --> `(?:foo)`
  (This could theoretically also have a positive performance impact, but
  I suspect JavaScript engines are already optimizing away capturing
  groups when they are not used.)

mary-poppins commented:
You can preview e3992f3 at https://pr32961-e3992f3.ngbuilds.io/.

jbogarthyde commented on packages/forms/src/validators.ts:
```diff
@@ -57,8 +57,36 @@ export const NG_VALIDATORS = new InjectionToken<Array<Validator|Function>>('NgVa
     new InjectionToken<Array<Validator|Function>>('NgAsyncValidators');
 
+/**
+ * At a high level, this regexp matches e-mail addresses of the format `local-part@tld`, where:
```
This should start with the context. Something like: 

A regular expression that is used by default to validate values in form controls that are configured to expect valid email values.

gkalpak replied:
I am not sure I agree :smiley:

Note that this is not part of public docs. This comment is only useful for people looking at the source code (which typically are project maintainers). Also, if someone is looking at this regexp trying to understand what it does, it is most likely that they ended up here _from_ the e-mail validator (so they know the context).

IMO, Source code comments on a pice of code should focus on what the piece of code does and not where/how it might be used elsewhere. Comments referring to other parts of the codebase (such as where this piece of code is used) are easy to get out-of-date.

Maybe just add something generic, such as:

> A regular expression that matches valid email addresses.

jbogarthyde commented on packages/forms/src/validators.ts:
```diff
@@ -57,8 +57,36 @@ export const NG_VALIDATORS = new InjectionToken<Array<Validator|Function>>('NgVa
+ *   punctuation symbols).
+ * - `local-part` cannot begin or end with a period (`.`).
+ * - `local-part` cannot be longer than 64 characters.
+ * - `tld` consistst of one or more `labels` separated by periods (`.`). E.g. `localhost` or
```
typo - consists. 
Also, please use "For example" or ", such as"  instead of e.g.

gkalpak replied:
:+1: (although keep in mind this is just a source code comment, not part of public docs)

jbogarthyde commented on packages/forms/src/validators.ts:
```diff
@@ -57,8 +57,36 @@ export const NG_VALIDATORS = new InjectionToken<Array<Validator|Function>>('NgVa
+ * ## Implementation background
+ *
+ * This regexp was ported over from AngularJS (see there for git history):
+ * https://github.com/angular/angular.js/blob/c133ef836/src/ng/directive/input.js#L27
```
how about:
 (see [git history](https://github.com/angular/angular.js/blob/c133ef836/src/ng/directive/input.js#L27)).

gkalpak replied:
Not sure :confused:
That link is not "git history". There are several ways to see the "git history", such as `git blame` or `git log` with special flags to track the evolution of a file or line of code. So, making `git history` a link to a source code line on GitHub is a little confusing.

(Note: If your intention was to avoid the lengthy link, keep in mind that this is a just a source code comment, so no HTML or Markdown links will work here anyway :smiley:)

jbogarthyde commented on packages/forms/src/validators.ts:
```diff
@@ -57,8 +57,36 @@ export const NG_VALIDATORS = new InjectionToken<Array<Validator|Function>>('NgVa
+ * https://github.com/angular/angular.js/blob/c133ef836/src/ng/directive/input.js#L27
+ * It is based on the
+ * [WHATWG version](https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address) with
+ * some enhancements to incorporate more RFC rules (e.g. related to domain names and the lengths of
```
such as (instead of e.g.)

gkalpak replied:
```suggestion
 * some enhancements to incorporate more RFC rules (such as rules related to domain names and the
```

jbogarthyde commented on packages/forms/src/validators.ts:
```diff
@@ -191,6 +219,20 @@ export class Validators {
+   * Tests the value using a [regular expression](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions)
+   * pattern suitable for common usecases. The pattern is based on the definition of a valid email
+   * address in the [WHATWG HTML specification](https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address)
+   * with some enhancements to incorporate more RFC rules (e.g. related to domain names and the
```
such as (instead of e.g.)

gkalpak replied:
```suggestion
   * with some enhancements to incorporate more RFC rules (such as rules related to domain names and
```

jbogarthyde reviewed (changes requested):
The doc itself looks good, but I'm concerned that most of it exists only in the file (although it is formatted for inclusion in AIO) -- There is the `Validators.email()` description, but  no reference in that to make anyone go look at the file for more information. 
Since the comments are formatted for processing by AIO, they are hard to read in the file..

gkalpak commented on packages/forms/src/validators.ts:
```diff
@@ -57,8 +57,36 @@ export const NG_VALIDATORS = new InjectionToken<Array<Validator|Function>>('NgVa
     new InjectionToken<Array<Validator|Function>>('NgAsyncValidators');
 
+/**
+ * At a high level, this regexp matches e-mail addresses of the format `local-part@tld`, where:
```
```suggestion
 * A regular expression that matches valid e-mail addresses.
 *
 * At a high level, this regexp matches e-mail addresses of the format `local-part@tld`, where:
```

gkalpak commented on packages/forms/src/validators.ts:
```diff
@@ -57,8 +57,36 @@ export const NG_VALIDATORS = new InjectionToken<Array<Validator|Function>>('NgVa
+ *   punctuation symbols).
+ * - `local-part` cannot begin or end with a period (`.`).
+ * - `local-part` cannot be longer than 64 characters.
+ * - `tld` consistst of one or more `labels` separated by periods (`.`). E.g. `localhost` or
```
```suggestion
 * - `tld` consists of one or more `labels` separated by periods (`.`). For example `localhost` or
```

gkalpak commented on packages/forms/src/validators.ts:
```diff
@@ -57,8 +57,36 @@ export const NG_VALIDATORS = new InjectionToken<Array<Validator|Function>>('NgVa
+ * It is based on the
+ * [WHATWG version](https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address) with
+ * some enhancements to incorporate more RFC rules (e.g. related to domain names and the lengths of
+ * different parts of the address). The main differences from the WHATWG version are:
```
```suggestion
 * lengths of different parts of the address). The main differences from the WHATWG version are:
```

gkalpak commented on packages/forms/src/validators.ts:
```diff
@@ -191,6 +219,20 @@ export class Validators {
+   * pattern suitable for common usecases. The pattern is based on the definition of a valid email
+   * address in the [WHATWG HTML specification](https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address)
+   * with some enhancements to incorporate more RFC rules (e.g. related to domain names and the
+   * lengths of different parts of the address).
```
```suggestion
   * the lengths of different parts of the address).
```

gkalpak committed:
fixup! docs(forms): expand e-mail validation description

mary-poppins commented:
You can preview 239ad95 at https://pr32961-239ad95.ngbuilds.io/.

gkalpak commented:
Thx for the review, @jbogarthyde! I addressed the comments (either by making the suggested changes or commenting back). PTAL

Also, my thoughts/questions regarding https://github.com/angular/angular/pull/32961#pullrequestreview-296360513:

> I'm concerned that most of it exists only in the file [...]

The main points are also mentioned in the `Validators.email()` description (which is part of the API docs). What info from the `EMAIL_REGEXP` comment you think would be useful in the API docs?

> There is the `Validators.email()` description, but no reference in that to make anyone go look at the file for more information.

We should generally not point people to the source code as a means of documentation. Looking at the source code is a great way to find out what the code does (although not well suited for less experienced developers), but when someone is looking at the documentation they probably want to read an explanation in prose. 

That said, for people that decide they do want to look at the source code (to find out more details), there is the "View source" button on the top right of the method description, which they can use to get to the `Validators.email()` implementation:

![view-method-source](https://user-images.githubusercontent.com/8604205/66196955-143ad080-e6a2-11e9-9cd3-cdcf21e4d534.png)

From there, getting to the `EMAIL_REGEXP` is really trivial (for someone intending to look at the source code).

> Since the comments are formatted for processing by AIO, they are hard to read in the file.

Not sure what you mean 😕
The comments are formatted in Markdown, which is supposed to be human-readable :wink: (It is also not specific to AIO - in fact that is how I format my comments in any project :grin:)
What exactly are you finding confusing? What would you format the comment for better readability?

AndrewKushnir commented:
Thanks for the changes @gkalpak! It looks like this PR needs a rebase, could you please help with rebase? I'll start presubmit after rebase as well, just to make sure RegExp change doesn't trigger any issues in g3. Thank you.

mary-poppins commented:
You can preview a651fae at https://pr32961-a651fae.ngbuilds.io/.

AndrewKushnir commented:
Thanks for rebasing this PR @gkalpak! I've started [VE presubmit](http://test/OCL:273109110:BASE:273109115:1570332283241:37758833) and also a [global VE run](https://test.corp.google.com/ui#id=OCL:273109110:BASE:273109163:1570332336756:c8250cf5) (to make sure RegExp change doesn't affect any targets). Will keep you updated. Thank you.

AndrewKushnir commented:
FYI, VE presubmits (including global one) are successful. Thank you.
